internal/workers/canary_manager.go

```diff
@@ -124,9 +125,14 @@ func (cm *CanaryManager) reconcile() {
 	glog.Infof("Canary manager reconcile ...")
 
 	if result, err := cm.topicService.Reconcile(); err == nil {
-		if result.RefreshMetadata {
+		if result.RefreshProducerMetadata {
 			cm.producerService.Refresh()
 		}
+
+		leaders, err := cm.consumerService.Leaders()
+		if err != nil || !reflect.DeepEqual(result.Leaders, leaders) {
```

Comment:
I'm not too happy with the allocation of the responsibilities between canary_manager.go and topic.go after this change.     In particular it is odd that RefreshProducerMetadata tells canary_manager.go to refresh the publisher metadata, but for consumer, the reconcile  has to work it out for itself.

I think it would be better if `TopicReconcileResult` was renamed `ReconcilationState`, and it carried the `[]*broker` returned from `DescribeCluster`.  That would give canary_manager's reconcile function sufficient knowledge to work out if either consumerService or producerService need their metadata refreshing. WDYT?

Reply:
> I think it would be better if TopicReconcileResult was renamed ReconcilationState, and it carried the []*broker returned from DescribeCluster. That would give canary_manager's reconcile function sufficient knowledge to work out if either consumerService or producerService need their metadata refreshing. WDYT?

It makes sense to me.

Reply:
I took a look at refactoring the responsibilities in the way described above.  It gets tricky because it would need to expose the initialisation state to the canary manager.  On reflection it actually makes the code worse rather than better.

So, other than add a unit testing (requires 207) I am actually proposing no further changes for this PR.